src/impl.ml

```diff
@@ -715,8 +715,18 @@ let write_stream common s destination _source_protocol destination_protocol prez
       (* TODO: https is not currently implemented *)
       let port = match Uri.port uri' with None -> (if use_ssl then 443 else 80) | Some port -> port in
       let host = match Uri.host uri' with None -> failwith "Please supply a host in the URI" | Some host -> host in
-      Lwt_unix.gethostbyname host >>= fun host_entry ->
-      let sockaddr = Lwt_unix.ADDR_INET(host_entry.Lwt_unix.h_addr_list.(0), port) in
+      let host =
+        match String.split_on_char '[' host with
+        | ""::[elem] ->
+          match String.split_on_char ']' elem with
+          | elem::[""] -> elem
+          | _ -> host
+        | _ -> host
+      in
+      Lwt_unix.getaddrinfo host (string_of_int port) [] >>= fun he ->
+      if List.length he = 0 then raise Not_found
+
+      let sockaddr = (List.hd he).Unix.ai_addr in
```

Comment:
So `gethostbyname` doesn't support ipv6 so i had to use `getaddrinfo`.
it returns an array of several socket with different protocols, should I prefer `TCP` or `UDP` here?

Reply:
We use `getaddrinfo` to determine FQDNs in xapi, so perhaps you could copy the implementation there?

https://github.com/xapi-project/xen-api/blob/372eb24c94671d7b756ab18111cb364e14a06e3e/ocaml/gencert/lib.ml#L22

Reply:
Yep it's more readable indeed.
Should I use CANONNAME as well or specific a portocol? TCP? UDP?

Reply:
Does taking the first reasonable looking `ai_canonname` from the list work? This is what xapi does, I think keeping the implementation the same as in xapi would be preferred (i.e. ignoring the protocol completely)

(NB: 
By 'first reasonable looking canonname', I mean the first canonname which satisfies 
```
           not (String.equal "" x
                || String.equal "localhost" x
                || Ipaddr.of_string x |> Stdlib.Result.is_ok)
```

As an aside, I would naively assume TCP is preferred over UDP (https://github.com/xapi-project/vhd-tool/blob/d889c029093f99e46f7107a9307e0bc5badc5c8b/src/impl.ml#L570 implies that UDP is not supported)
)

Reply:
The previous codes used IP addresses, using canonname would change the behavior.
I'm not sure it's wanted since, we use the address to create a socket.

The protocol is not used in the rest of the code. So my previous question is useless.

Reply:
Well the old code just did `host_entry.Lwt_unix.h_addr_list.(0)`, so I think taking head of the list is fine here (i.e. the new code you have written looks right to me)

Reply:
I agree